### PR TITLE
Fix functions execution over sparse columns

### LINKED ISSUE #55047: isDefaultAt is not implemented for Function
v.23.8.2.7

I run the same query on the same data, and sometimes it throws an error, and sometimes it works without problems.

` Code: 48. DB::Exception: isDefaultAt is not implemented for Function: while executing 'FUNCTION Capture[Int64](Int64) -> UInt8(toInt64(prev_id) :: 0) -> __lambda Function(Int64 -> UInt8) : 3'. (NOT_IMPLEMENTED) (version 23.8.2.7 (official build))`

```
drop table if exists test_table;
create table test_table engine MergeTree order by id as
select assumeNotNull(id)              as id,
       arrayCount((cr_id1) -> (toInt64(cr_id1) = toInt64(t.prev_id)),
                   t2.ids) as cnt
from mytable t
         left join (select client_id,
                           groupArray(id)                                   as ids
                    from mytable
                    group by client_id) t2
                   on t.client_id = t2.client_id
limit 100;
```

### PR DESCRIPTION
**Pull request by azat**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix functions execution over sparse columns (fixes `DB::Exception: isDefaultAt is not implemented for Function: while executing 'FUNCTION Capture` error)

Cc: @CurtizJ 
Fixes: #55047

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit ce7f0e5ae781889cde1e621a35d07e2e38681611 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/55275/ce7f0e5ae781889cde1e621a35d07e2e38681611/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6432807101">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/55275/ce7f0e5ae781889cde1e621a35d07e2e38681611/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Comment by azat** [comment]
>Integration tests (tsan) [1/6] — fail: 1, passed: 390
>Stress test (tsan) — check_status.tsv doesn't exists
>Stateful tests (tsan) — Invalid check_status.tsv
>Stateful tests (tsan, ParallelReplicas) — Invalid check_status.tsv

- #55260 

>Stateful tests (release, ParallelReplicas) — Invalid check_status.tsv

```
Traceback (most recent call last):
  File "/s3downloader", line 114, in <module>

    unpack_to_clickhouse_directory(temp_archive_path, args.clickhouse_data_path)
  File "/s3downloader", line 85, in unpack_to_clickhouse_directory
    comp_file.extractall(path=clickhouse_path)
  File "/usr/lib/python3.10/tarfile.py", line 2257, in extractall
    self._extract_one(tarinfo, path, set_attrs=not tarinfo.isdir(),
  File "/usr/lib/python3.10/tarfile.py", line 2320, in _extract_one
    self._extract_member(tarinfo, os.path.join(path, tarinfo.name),
  File "/usr/lib/python3.10/tarfile.py", line 2403, in _extract_member
    self.makefile(tarinfo, targetpath)
  File "/usr/lib/python3.10/tarfile.py", line 2456, in makefile
    copyfileobj(source, target, tarinfo.size, ReadError, bufsize)
  File "/usr/lib/python3.10/tarfile.py", line 254, in copyfileobj
    raise exception("unexpected end of data")
tarfile.ReadError: unexpected end of data
```

File: src/Functions/IFunction.cpp [hunk 1]
```diff
@@ -310,7 +311,10 @@ ColumnPtr IExecutableFunction::executeWithoutSparseColumns(const ColumnsWithType
 
 ColumnPtr IExecutableFunction::execute(const ColumnsWithTypeAndName & arguments, const DataTypePtr & result_type, size_t input_rows_count, bool dry_run) const
 {
-    if (useDefaultImplementationForSparseColumns())
+    bool use_default_implementation_for_sparse_columns = useDefaultImplementationForSparseColumns();
+    /// DataTypeFunction does not supports obtaining default (isDefaultAt())
```
Comment (alexey-milovidov) [inline src/Functions/IFunction.cpp:315]: Why don't return a default function, something like ignore?

```
milovidov-desktop :) SELECT ignore()

SELECT ignore()

Query id: fd69092d-62be-42ab-8c63-752def7e7dd1

┌─ignore()─┐
│        0 │
└──────────┘
```

File: src/Functions/IFunction.cpp [hunk 2]
```diff
@@ -310,7 +311,10 @@ ColumnPtr IExecutableFunction::executeWithoutSparseColumns(const ColumnsWithType
...
+    /// ColumnFunction does not uspport getting specific values
```
Comment (alexey-milovidov) [inline src/Functions/IFunction.cpp:316]: A typo: `uspport`

Why don't return a tuple of captured arguments as a value?

File: src/Functions/IFunction.cpp (hunk 1 above)
Comment (azat) [reply src/Functions/IFunction.cpp:315]: It is not interchangeable - https://github.com/ClickHouse/ClickHouse/blob/b38d4b5b0fb6956242e9a870ea5d6275210f3a22/src/Functions/IFunction.cpp#L361-L368

File: src/Functions/IFunction.cpp (hunk 2 above)
Comment (azat) [reply src/Functions/IFunction.cpp:316]: I don't see how it is interchangeable - https://github.com/ClickHouse/ClickHouse/blob/b38d4b5b0fb6956242e9a870ea5d6275210f3a22/src/Functions/IFunction.cpp#L361-L368

`canBeInsideSparseColumns` can be used here, but there is special handling of const result, so I decided not to change this.

**Review by alexey-milovidov — APPROVED** [review]

**Pull request merged**